### DISCUSSION TITLE
Why is a PR created which bumps the version when `master` is same as latest release tag?

### DISCUSSION THREAD

**Discussion by mightyiam** [release-plz/release-plz#1105] [Q&A] [answered]
I realize there are other considerations, but I just don't understand them.

https://github.com/mobusoperandi/eelco/pull/17#issuecomment-1837524112

**Comment by marcoieni** [comment]
Te changelog says "updated dependencies". Maybe you updated cargo.lock or cargo.toml? I'm not at the laptop right now, so I can't check.

**Reply by mightyiam** [reply]
I'm not sure how to answer this, because there is only one release.

**Comment by marcoieni** [comment, marked as answer]
The issue is that release-plz is detecting a change in the lock file due to [this](https://github.com/mobusoperandi/eelco/commit/7ff66b7a88a64343db15945e2a59342205fdd399) commit.

Cargo.lock diff:
![image](https://github.com/MarcoIeni/release-plz/assets/11428655/456bf20d-6225-4b08-b7a9-1f0c8a97466c)

**Reply by mightyiam** [reply]
Ahm... thank you for the answer. But, sorry, I'm still confused. That commit is before the commit that the latest release tag points to. In other words, that commit is before the latest release. That is, according to git tags, it is. But perhaps release-plz uses some other logic to determine what the last release is and therefore what commits are considered for next release changelog and version bump?

**Reply by marcoieni** [reply]
> perhaps release-plz uses some other logic to determine what the last release is 

Yes, it compares the local package with the package in crates.io. 

If you run `cargo package --target-dir my-dir` and than do `cat my-dir/package/eelco-0.1.0/Cargo.lock` you will find that cargo is ignoring your `Cargo.lock` git override.
So the Cargo.lock you have published to `crates.io` is different then your local one.
Consider doing an override in the Cargo.toml if possible: https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html

**Reply by mightyiam** [reply]
Wow. I find that somewhat counterintuitive, if I understand at all. Will study during the weekend. Thank you, @MarcoIeni.

**Reply by mightyiam** [reply]
There might be a problem with implementation. We found that crates.io never accepts non-crates.io dependencies (bottom of [this section](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories)). And we didn't know this. So, this changes things for us a little bit, because we'd have to either not publish to crates.io, which happens to be okay for us, because our users would be consuming the crate via Nix, in which crates.io is optional.

But for release-plz, this realization may mean that it's currently doing something incorrect. It seems to be comparing the published Cargo.lock with the local Cargo.lock, isn't it? If that's true, the problem with that is that they are inherently different files. Specifically, the published Cargo.lock is cleaned of non-crates.io sources, such as git sources, while the local, checked-in Cargo.lock has them. What the published Cargo.lock would have are all crates.io sources that satisfy the version range, instead of git sources.

We suppose what could be a possible solution in the scope of the current release-plz design, is to disregard the checked-in Cargo.lock and compare against a Cargo.lock that would be obtained via `cargo package`. In that case, the PR in the original question would not have been created, because there would have been no difference between the last published Cargo.lock and the current "package" Cargo.lock.

**Reply by marcoieni** [reply]
I implemented https://github.com/MarcoIeni/release-plz/pull/1141 to fix your issue. I tested `release-plz update` on eelco and now release-plz doesn't detect a new change 👍

**Reply by mightyiam** [reply]
Thank you!
